### PR TITLE
Layout gap value list support

### PR DESCRIPTION
**Pull request by lievenjanssen**
Some additional improvements we could do:
- Support $increment in the gap mixins (e.g. `gap-class-selectors($from: 4, $to: 24, $increment: 4, $gapUnits: px em rem, $includeSelectorsForMediaSizes: false)` but that would be a breaking change.
- Add flex-row-properties to elements that don't have fx-layout and have fx-gap*, like we do for fx-flex: https://github.com/philmtd/css-fx-layout/blob/c0b3b022ab1ddf91181386ea8aa8f823157c1589/src/lib/flex-classes.scss#L13-L15 We could also do this for fx-align*.

What do you think?

### CODE REVIEW

File: _index.scss [hunk 1]
```diff
@@ -2,6 +2,3 @@
 	gap-class-selectors,
 	attribute-selectors,
 	gap-attribute-selectors;
-
-@forward "./src/lib/mixins" hide layout-gap-attributes,
-	layout-gap-classes;
```
Comment (philmtd) [inline _index.scss:7]: This removes all mixins from the public API. We should keep the following in the file:
```scss
@forward "./src/lib/mixins";
```

File: src/lib/layout-gap-attributes.scss [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+///
+/// @param {number} $from
+///   Will start generating selectors from $from $gapUnits, $from can also be used as a value list to generate value $gapUnits for every value in $from
+/// @param {number} $to
+///   Will generate selectors until including $to $gapUnits
+/// @param {string} $gapUnits
+///   Will generate selectors for the units passed in $gapUnits, e.g. px em rem
+/// @param {bool} $includeSelectorsForMediaSizes
+///	  Will generate additional selectors for flex layout media sizes
+@mixin layout-gap-attributes($from: 1, $to: 24, $gapUnits: px em rem, $includeSelectorsForMediaSizes: true) {
```
Comment (philmtd) [inline src/lib/layout-gap-attributes.scss:31]: To be honest I think the double functionality of the `$from` parameter is difficult, although it keeps the API unchanged while extending the functionality. What do you think about adding a second mixin?

**Comment by philmtd** [comment]
Thank you very much for your contribution.

> Support $increment in the gap mixins (e.g. gap-class-selectors($from: 4, $to: 24, $increment: 4, $gapUnits: px em rem, $includeSelectorsForMediaSizes: false) but that would be a breaking change.

Couldn't we add $increment as the last parameter with a default value of 1 in the "$from-$to"-mixin? That way we would not implement a breaking change, right?

Regarding my second comment: What do you think about using separate mixins? One for "$from-$to" and another for a fixed list of values?

> Add flex-row-properties to elements that don't have fx-layout and have fx-gap*, like we do for fx-flex

> We could also do this for fx-align*.

Sure, feel free to add that :)

**Review by philmtd — CHANGES_REQUESTED** [review]

File: _index.scss (hunk 1 above)
Comment (lievenjanssen) [reply _index.scss:7]: I'll add it again.

File: src/lib/layout-gap-attributes.scss (hunk 2 above)
Comment (lievenjanssen) [reply src/lib/layout-gap-attributes.scss:31]: Agreed. What do you suggest to name the list mixins?

**Comment by lievenjanssen** [comment]
`Couldn't we add $increment as the last parameter with a default value of 1 in the "$from-$to"-mixin? That way we would not implement a breaking change, right?`

I feel that logically $increment should be a parameter before $includeSelectorsForMediaSizes but I agree we could add it as the last parameter so we don't need to introduce a breaking change. Will also move $increment parameter to the last position in the flex size mixins.

**Comment by lievenjanssen** [comment]
I'll also add the automatic row layout for layout gap and layout align classes and attributes.

File: src/lib/layout-gap-attributes.scss (hunk 2 above)
Comment (philmtd) [reply src/lib/layout-gap-attributes.scss:31]: Maybe something as simple as `layout-gap-attributes-list` or variations of this with `-as-list` or `-fixed-list` or `-with-values`?

I think I like the `-as-list` suffix the most. What do you think?

**Comment by philmtd** [comment]
> `Couldn't we add $increment as the last parameter with a default value of 1 in the "$from-$to"-mixin? That way we would not implement a breaking change, right?`
> 
> I feel that logically $increment should be a parameter before $includeSelectorsForMediaSizes but I agree we could add it as the last parameter so we don't need to introduce a breaking change. Will also move $increment parameter to the last position in the flex size mixins.

Yes, the logical order of the parameters is not the best then. An alternative would be to add another mixin.. but I think this may overcomplicate the API. :/

File: src/lib/layout-gap-attributes.scss (hunk 2 above)
Comment (philmtd) [reply src/lib/layout-gap-attributes.scss:31]: Yes, that's also good.

**Comment by lievenjanssen** [comment]
> > `Couldn't we add $increment as the last parameter with a default value of 1 in the "$from-$to"-mixin? That way we would not implement a breaking change, right?`
> > I feel that logically $increment should be a parameter before $includeSelectorsForMediaSizes but I agree we could add it as the last parameter so we don't need to introduce a breaking change. Will also move $increment parameter to the last position in the flex size mixins.
> 
> Yes, the logical order of the parameters is not the best then. An alternative would be to add another mixin.. but I think this may overcomplicate the API. :/

Maybe we should introduce a breaking change? I think having the ability to increment the gaps makes total sense.

File: src/lib/layout-gap-attributes.scss (hunk 2 above)
Comment (lievenjanssen) [reply src/lib/layout-gap-attributes.scss:31]: Ok I'll do that.

**Comment by philmtd** [comment]
> Maybe we should introduce a breaking change? I think having the ability to increment the gaps makes total sense.

I guess we're in the lucky situation that this library is not that popular (yet?), so we would not break many dependent projects.

To make sure the mixins are invoked with correct argument types, do you think we could/should add type assertions? I'm not familiar with the best practices in SCSS but I was thinking about something similar to this in the mixin:

```scss
@if (type-of($to) != "number") {
    @error("$to should be a number");
}
```

**Comment by lievenjanssen** [comment]
> I guess we're in the lucky situation that this library is not that popular (yet?), so we would not break many dependent projects. So yes, I'd be open to introduce this as a breaking change. I mean.. I could just release it as a new major version to make things obvious. Or just leave it as the next minor because it does not affect too many projects. What are your thoughts on it?

If it's a breaking change I would definitely go for a new major version. There's going to be quite some changes in it when we also include the flex-size changes so a new major version makes sense.

**Comment by lievenjanssen** [comment]
> To make sure the mixins are invoked with correct argument types, do you think we could/should add type assertions?

Not sure if it's really needed. I think it's quite obvious. I added a check for $increment for flex-sizes though (will still open the PR later today) since both $from and $to need to be dividable by $increment in order to make the logic work.

**Comment by lievenjanssen** [comment]
For me this can be merged now.

I'll open seperate PR's for the increment change and automatic row layout for layout gap and layout align classes and attributes.

**Review by philmtd — APPROVED** [review]

**Pull request merged**